### PR TITLE
Add Geofence to Mission UI 

### PR DESCRIPTION
**Pull request by joesi8302**
## Why is the change being made?
This change was being made to add Geofencing fields to mission window.
 ## What has changed to address the problem?
This change added a bounding box to the Mission UI to simulate a geofence.
 ## How was this change tested?
This change was tested with `npm test`

### CODE REVIEW

**Comment by Luis729** [comment]
@joesi8302 set me as reviewer, not assignee

**Comment by Luis729** [comment]
Also a few issues before I review code:
- [x] Can you set the title of the commit & PR to `Adding Geofence to Mission UI`, remove the parenthesized part
- [x] Can you rebase master into this branch to update it? https://ngcp-calpoly.quip.com/0aJOAkm70ZzI/NGCP-Coding-Standards-and-Practices#MTQACAZZmyZ

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 1]
```diff
@@ -4,15 +4,18 @@ import { Event, ipcRenderer } from 'electron';
 import React, { Component, ReactNode } from 'react';
 import * as Slider from 'rc-slider';
 
+import { BoundingBoxBounds, ThemeProps } from '../../types/componentStyle';
+
 import { JobType } from '../../static/index';
 
-import { ThemeProps } from '../../types/componentStyle';
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: move line 7 back here

Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx:9]: move line 7 back here

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 2]
```diff
@@ -4,15 +4,18 @@ import { Event, ipcRenderer } from 'electron';
...
 import * as MissionInformation from '../../types/missionInformation';
 import { VehicleObject, VehicleStatus } from '../../types/vehicle';
 
 import ipc from '../../util/ipc';
 import { updateVehicles } from '../../util/util';
 
+import { readyToStart } from '../../util/parameter';
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: remove the empty line in line 16

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 3]
```diff
@@ -22,6 +25,8 @@ import UGVRescue from './parameter/UGVRescue';
 import UUVRescue from './parameter/UUVRescue';
 import VTOLSearch from './parameter/VTOLSearch';
 
+import CreateBoundingBoxButton from './extra/CreateBoundingBoxButton';
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: go ahead and put this between line 19 and 20

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 4]
```diff
@@ -53,8 +58,40 @@ const title: { [missionName in MissionInformation.MissionName]: string } = {
   ugvRescue: 'UGV Rescue',
   uuvRescue: 'UUV Rescue',
 };
+//  Geofencing implementation
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: no need for this comment

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 5]
```diff
@@ -53,8 +58,40 @@ const title: { [missionName in MissionInformation.MissionName]: string } = {
...
+const checklistCache: { [check in GeofenceChecklistType ]: number | undefined} = {
+  geofenceTop: undefined,
+  geofenceLeft: undefined,
+  geofenceRight: undefined,
+  geofenceBottom: undefined,
+};
+
+type GeofenceType = 'geofence';
+
+type Locked = { [type in GeofenceType]: boolean} & {
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: simplify line 71 and 73-75 by the following:
```ts
type Locked = {
  geofence: boolean;
};
```

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 6]
```diff
@@ -155,9 +193,18 @@ export default class MissionWindow extends Component<ThemeProps, State> {
     this.stopMissions = this.stopMissions.bind(this);
     this.toggleMissionType = this.toggleMissionType.bind(this);
     this.tipFormatter = this.tipFormatter.bind(this);
+
+    this.onChange = this.onChange.bind(this);
+    this.updateVehicles = this.updateVehicles.bind(this);
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: why did you move line 148 to here

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 7]
```diff
@@ -175,6 +222,76 @@ export default class MissionWindow extends Component<ThemeProps, State> {
       endMissionIndex: value[1],
     });
   }
+  // GeoFence BoundingBox
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: remove this comment

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 8]
```diff
@@ -175,6 +222,76 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
+  private onChange(event: React.ChangeEvent<HTMLInputElement>): void {
+    const { checklist } = this.state;
+    if (!(event.target.name in checklist)) return;
+
+    const name = event.target.name as GeofenceChecklistType;
+    const value = parseInt(event.target.value, 10) || 0;
+    switch (name) {
+      case 'geofenceTop':
+        ipc.postUpdateBoundingBoxes(true, {
+          name: 'Bounding Box',
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: im pretty sure the name wouldnt be Bounding Box, probably should be Geofencing or Geofencing Box

make sure to update the other three cases too and change Bounding Box to your geofencing name

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 9]
```diff
@@ -201,12 +318,54 @@ export default class MissionWindow extends Component<ThemeProps, State> {
     this.setState({ activeVehicleMapping: newActiveVehicleMapping });
   }
 
-  private updateInformation(information: MissionInformation.Information): void {
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: why was this moved? this used to be under `updateVehicles()`

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 10]
```diff
@@ -201,12 +318,54 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
-    const { information: currentInformation } = this.state;
-    const newInformation = currentInformation;
+  private postStartMissions(): void {
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: same here, why are all these functions moved around?

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 11]
```diff
@@ -260,60 +419,53 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
-    this.setState({ status: 'running' });
-  }
+  private updateBoundingBoxes(
+    ...boundingBoxes: { name: string; bounds: BoundingBoxBounds }[]
+  ): void {
+    const checks: { [checkName in GeofenceChecklistType]?: number} = {};
+
+    boundingBoxes.forEach((boxpoint): void => {
+      switch (boxpoint.name) {
+        case 'Bounding Box':
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: shouldnt be Bounding Box

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 12]
```diff
@@ -366,6 +517,17 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
+        <div className="geofenceContainer">
+          <h1> GeoFencing </h1>
+          <input className="inputFields" type="number" name="geofenceScanTop" value={checklist.geofenceTop || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Top" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceBottom" value={checklist.geofenceBottom || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Bottom" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceLeft" value={checklist.geofenceLeft || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Left" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceRight" value={checklist.geofenceRight || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Right" />
+          <CreateBoundingBoxButton theme={theme} name="geofence" value="Bounding Box" />
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: shouldnt be Bounding Box

**Review by Luis729 — CHANGES_REQUESTED** [review]

File: src/renderer/missionWindow/mission.css [hunk 13]
```diff
@@ -8,9 +8,12 @@
   grid-template-areas:
     'selector selector selector mapping'
     'parameter parameter parameter mapping'
-    'parameter parameter parameter options'
+    'parameter parameter parameter mapping'
+    'parameter parameter parameter mapping'
+    'geofence geofence geofence  options'
```
Comment (ckid199) [inline src/renderer/missionWindow/mission.css]: Extra space in between geofence and options for both lines 13 and 14

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 14]
```diff
@@ -366,6 +517,17 @@ export default class MissionWindow extends Component<ThemeProps, State> {
           <h1 style={{ marginTop: 0 }}>Options</h1>
           <MissionOptions title={title} missionNames={missionNames} options={options} />
         </div>
+        <div className="geofenceContainer">
+          <h1> GeoFencing </h1>
```
Comment (ckid199) [inline src/renderer/missionWindow/MissionWindow.tsx]: Change the capitalization to Geofencing

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 9 above)
Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx]: When i was doing `npm test` it kept telling me to move them around. Not sure why

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 1 above)
Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx:9]: Where to exactly? the import { JobType } goes to line 7 and the import { BoundingBoxBounds, ThemeProps } goes where?

**Review by ckid199 — COMMENTED** [review]
Noticed some odd capitalization

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 1 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx:9]: See the image below to see what I mean
image

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 9 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: Oh I see, that's fine then. I'll look at the code afterwards

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 12 above)
Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx]: just the value="Bounding Box" right?

Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx]: if i would change every function with the word "boundingbox" would it break the code?

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 9 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: You can actually keep `updateInformation()` after `updateVehicles()` and it should be fine

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 10 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: yeah you actually dont need to move any of these functions around, i checked it out for myself just now. (in other words put the functions back)

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 12 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: so when you have value="Bounding Box" its calling that specific box "Bounding Box", you'd want to change it to "Geofencing" or "Geofencing Box". This is completely different from the name="geofence" field

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 14 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: also remove the space from the h1 tag

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 15]
```diff
@@ -6,16 +6,18 @@ import * as Slider from 'rc-slider';
 
 import { JobType } from '../../static/index';
 
-import { ThemeProps } from '../../types/componentStyle';
+import { BoundingBoxBounds, ThemeProps } from '../../types/componentStyle';
+
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: Remove this line

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 16]
```diff
@@ -6,16 +6,18 @@ import * as Slider from 'rc-slider';
...
 
 import ipc from '../../util/ipc';
 import { updateVehicles } from '../../util/util';
+import { readyToStart } from '../../util/parameter';
 
 import ActiveVehicleMapping from './extra/ActiveVehicleMapping';
 import MissionOptions from './extra/MissionOptions';
-
+import CreateBoundingBoxButton from './extra/CreateBoundingBoxButton';
 import ISRSearch from './parameter/ISRSearch';
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx:19]: Put empty line between line 20 and 21

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 6 above)
Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx]: image

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 17]
```diff
@@ -271,7 +405,7 @@ export default class MissionWindow extends Component<ThemeProps, State> {
       startMissionIndex,
     } = this.state;
 
-    // Gets all relevant mission information for the missions being performed for the mission type.
+    //  Gets all relevant mission information for the missions being performed for the mission type.
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: Remove extra space

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 18]
```diff
@@ -339,7 +483,6 @@ export default class MissionWindow extends Component<ThemeProps, State> {
      * Start button will not appear unless all mission information is filled out
      * (and no mission is running).
      */
-    const readyToStart = information[missionName].parameters !== undefined;
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: why was this deleted

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 19]
```diff
@@ -310,10 +444,20 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
+    }
+
+    this.setState({ locked: newLocked });
+  }
+
   public render(): ReactNode {
+    const { checklist, locked } = this.state;
     const { theme } = this.props;
     const {
-      information,
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx:316]: why was this deleted

**Review by Luis729 — CHANGES_REQUESTED** [review]

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 20]
```diff
@@ -366,6 +509,17 @@ export default class MissionWindow extends Component<ThemeProps, State> {
...
+        <div className="geofenceContainer">
+          <h1> Geofencing </h1>
+          <input className="inputFields" type="number" name="geofenceScanTop" value={checklist.geofenceTop || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Top" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceBottom" value={checklist.geofenceBottom || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Bottom" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceLeft" value={checklist.geofenceLeft || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Left" />
+          <br />
+          <input className="inputFields" type="number" name="geofenceRight" value={checklist.geofenceRight || ''} disabled={locked.geofence} onChange={this.onChange} placeholder="Right" />
+          <CreateBoundingBoxButton theme={theme} name="geofence" value="GeoFencing Box" />
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: Can you rename all "GeoFencing Box" to "Geofencing Box" or "Geofencing"

**Review by Luis729 — CHANGES_REQUESTED** [review]

File: src/renderer/missionWindow/MissionWindow.tsx (hunk 19 above)
Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: from what i remember, it gave me an error saying it was not being used.

Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: So I think it is because you imported a function called `readyToStart` (line 14) but the variable `readyToStart` checks for something else. I think this is why you deleted it.

I recommend you rename the function when you import it:

```ts
import { readyToStart as giveMeANameBoss } from '../../util/parameter';
```

or rename the variable:

```ts
const giveMeANameBoss = information[missionName].parameters !== undefined;
```

Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: update, the variable should have the new logic (if you chose to rename the variable the example is as shown):

```ts
const giveMeANameBoss = information[missionName].parameters !== undefined && readyToStart(this);
```

Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: Lets assume for the sake of this PR that we choose to rename the variable to `unlockStartMissionButton`. It is a very clear name for the variable and what it does (unlock the start mission button so that it can be clicked).

```ts
const unlockStartMissionButton = information[missionName].parameters !== undefined;
```

So in order to update `unlockStartMissionButton` to include requiring all geofencing information to be filled out too, the variable needs to be updated as such:

```ts
const unlockStartMissionButton = information[missionName].parameters !== undefined
  && checklist.geofenceTop !== undefined
  && checklist.geofenceBottom !== undefined
  && checklist.geofenceLeft !== undefined
  && checklist.geofenceRight !== undefined;
```

## There is no need to import the `readyToStart` function from `parameter.ts`, but keep reading if you would like to know how to integrate this function too:

The `readyToStart` function's job is to check if all the values in the state variable `checklist` is not `undefined`. Then if all the values in the `checklist` is not `undefined`, it will set another state variable `ready` to true. The logic for `unlockStartMissionButton` can be updated as shown:

```ts
const { checklist, locked, ready } = this.state;

const unlockStartMissionButton = information[missionName].parameters !== undefined &&
  (ready || readyToStart(this));
```

When `readyToStart` actually returns true, the `ready` variable is updated to `true`. After that, the `unlockStartMissionButton` will never run `readyToStart` since `ready === true`.

Talking about Big-O complexity, the speed that `readyToStart` runs at is O(N). The speed after `ready` is true changes it to O(1).

Comment (joesi8302) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: so the code without the `readyToStart` function seems to work, however if i were to use the code with the function, i would get similar errors as before.

Comment (Luis729) [reply src/renderer/missionWindow/MissionWindow.tsx:316]: I dont think its a big deal with the function so feel free to leave it out. It only has 4 additional checks so 5 checks vs 1 check shouldnt be a big performance thing. Feel free to stick to the non-function one.

Im looking at the function right now myself and I dont really like it that much - I dont really like the entire mission UI class and it should be made better in the future so I'm sure that function will be removed in the short future.

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 21]
```diff
@@ -5,16 +5,15 @@ import React, { Component, ReactNode } from 'react';
...
 import { JobType } from '../../static/index';
-
-import { ThemeProps } from '../../types/componentStyle';
+import { BoundingBoxBounds, ThemeProps } from '../../types/componentStyle';
 import * as MissionInformation from '../../types/missionInformation';
 import { VehicleObject, VehicleStatus } from '../../types/vehicle';
 
 import ipc from '../../util/ipc';
 import { updateVehicles } from '../../util/util';
-
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: lets keep an extra empty line here

File: src/renderer/missionWindow/MissionWindow.tsx [hunk 22]
```diff
@@ -339,7 +483,12 @@ export default class MissionWindow extends Component<ThemeProps, State> {
      * Start button will not appear unless all mission information is filled out
      * (and no mission is running).
      */
-    const readyToStart = information[missionName].parameters !== undefined;
+
+    const unlockStartMissionButton = information[missionName].parameters !== undefined
+    && checklist.geofenceTop !== undefined
```
Comment (Luis729) [inline src/renderer/missionWindow/MissionWindow.tsx]: lets indent these && statements by two

ex:
```ts
const a = something
  && somethingElse;
```

**Review by Luis729 — CHANGES_REQUESTED** [review]

**Review by Luis729 — APPROVED** [review]

**Comment by Luis729** [comment]
Looks awesome :shipit:
